In [1]:
pip install imbalanced-learn

     |████████████████████████████████| 167 kB 4.3 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd
url="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_fa50cdf1f0d84f8f95a128ef173ba668.csv"
df=pd.read_csv(url)
df.head()

behavior_sexualRisk  behavior_eating  behavior_personalHygine  \
0                   10               13                       12   
1                   10               11                       11   
2                   10               15                        3   
3                   10               11                       10   
4                    8               11                        7   

   intention_aggregation  intention_commitment  attitude_consistency  \
0                      4                     7                     9   
1                     10                    14                     7   
2                      2                    14                     8   
3                     10                    15                     7   
4                      8                    10                     7   

   attitude_spontaneity  norm_significantPerson  norm_fulfillment  \
0                    10                       1                 8   
1                     7                       5                 5   
2                    10                       1                 4   
3                     7                       1                 5   
4                     8                       1                 5   

   perception_vulnerability  perception_severity  motivation_strength  \
0                         7                    3                   14   
1                         4                    2                   15   
2                         7                    2                    7   
3                         4                    2                   15   
4                         3                    2                   15   

   motivation_willingness  socialSupport_emotionality  \
0                       8                           5   
1                      13                           7   
2                       3                           3   
3                      13                           7   
4                       5                           3   

   socialSupport_appreciation  socialSupport_instrumental  \
0                           7                          12   
1                           6                           5   
2                           6                          11   
3                           4                           4   
4                           6                          12   

   empowerment_knowledge  empowerment_abilities  empowerment_desires  \
0                     12                     11                    8   
1                      5                      4                    4   
2                      3                      3                   15   
3                      4                      4                    4   
4                      5                      4                    7   

   ca_cervix  
0          1  
1          1  
2          1  
3          1  
4          1

In [2]:
df.shape

(72, 20)

In [3]:
df.groupby("ca_cervix")["ca_cervix"].count()

ca_cervix
0    51
1    21
Name: ca_cervix, dtype: int64

In [4]:
X=df.drop(["ca_cervix"],axis=1)
y=df[["ca_cervix"]]

In [5]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X=pd.DataFrame(scaler.fit_transform(X),columns=X.columns)
X.head()

behavior_sexualRisk  behavior_eating  behavior_personalHygine  \
0             0.282843         0.088848                 0.304267   
1             0.282843        -0.764090                -0.027661   
2             0.282843         0.941785                -2.683082   
3             0.282843        -0.764090                -0.359588   
4            -1.414214        -0.764090                -1.355371   

   intention_aggregation  intention_commitment  attitude_consistency  \
0              -1.435337             -2.691824              1.203152   
1               0.771302              0.276840             -0.119397   
2              -2.170883              0.276840              0.541878   
3               0.771302              0.700934             -0.119397   
4               0.035756             -1.419539             -0.119397   

   attitude_spontaneity  norm_significantPerson  norm_fulfillment  \
0              0.922767               -1.159390         -0.099748   
1             -1.070410                1.022992         -0.715338   
2              0.922767               -1.159390         -0.920534   
3             -1.070410               -1.159390         -0.715338   
4             -0.406017               -1.159390         -0.715338   

   perception_vulnerability  perception_severity  motivation_strength  \
0                 -0.356554            -0.707394             0.423008   
1                 -1.063120            -1.003512             0.736994   
2                 -0.356554            -1.003512            -1.774891   
3                 -1.063120            -1.003512             0.736994   
4                 -1.298641            -1.003512             0.736994   

   motivation_willingness  socialSupport_emotionality  \
0               -0.413116                   -0.735053   
1                0.805914                   -0.260400   
2               -1.632146                   -1.209707   
3                0.805914                   -0.260400   
4               -1.144534                   -1.209707   

   socialSupport_appreciation  socialSupport_instrumental  \
0                    0.289642                    0.379106   
1                   -0.057928                   -1.253965   
2                   -0.057928                    0.145810   
3                   -0.753070                   -1.487260   
4                   -0.057928                    0.379106   

   empowerment_knowledge  empowerment_abilities  empowerment_desires  
0               0.336305               0.404687            -0.511741  
1              -1.277960              -1.280951            -1.410408  
2              -1.739179              -1.521756             1.060926  
3              -1.508570              -1.280951            -1.410408  
4              -1.277960              -1.280951            -0.736408

In [6]:
import numpy as np
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

# feature extraction. We are going to output the selection scores for all features and select the features with the highest scores.
test = SelectKBest(score_func=f_classif, k="all")
fit = test.fit(X, np.asarray(y).reshape(y.shape[0],))
features = fit.transform(X)
# summarize scores
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

scores
17  29.171896
10  24.375813
16  21.163032
11  19.558959
18  18.664926
12  15.787657
8   15.669342
9   15.264080
13  12.702173
2   11.294690
14   7.619211
0    7.502791
7    6.614904
3    6.511008
4    4.167306
1    2.927568
15   1.290717
5    1.119208
6    0.504491

In [8]:
#X_score=scores[scores["scores"]>=10]
#len(X_score)

In [89]:
#X=X.iloc[:,X_score.index]
#X.head()

In [90]:
X.shape

(72, 19)

In [9]:
from imblearn.ensemble import BalancedRandomForestClassifier
from sklearn import model_selection
from sklearn.model_selection import KFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [10]:
BForest=BalancedRandomForestClassifier(n_estimators=200,criterion="entropy",max_depth=6,min_samples_split=3,
                                       min_samples_leaf=2,max_features="sqrt",bootstrap=True,
                                       sampling_strategy="all",replacement=False,n_jobs=-1,warm_start=False,
                                       random_state=42)
scores=cross_val_score(BForest,X,np.asarray(y).reshape(y.shape[0],),cv=13)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.9179487179487181 +/- 0.08926612915656004


In [11]:
from sklearn.metrics import roc_auc_score
y_pred = cross_val_predict(BForest, X, np.asarray(y).reshape(y.shape[0],), cv=13)
AUC = roc_auc_score(y, y_pred)
print("AUC: ",AUC)

AUC:  0.9131652661064426


In [12]:
from sklearn.metrics import confusion_matrix
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[47  4]
 [ 2 19]]


In [64]:
from imblearn.ensemble import BalancedBaggingClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn import svm
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.gaussian_process import GaussianProcessClassifier
BBagging=BalancedBaggingClassifier(base_estimator=GaussianNB(),
                                    n_estimators=22,bootstrap_features=False,sampling_strategy="not majority",
                                   oob_score=False,n_jobs=-1,random_state=0)
scores=cross_val_score(BBagging,X,np.asarray(y).reshape(y.shape[0],),cv=8)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.9305555555555556 +/- 0.053791435363991925


In [65]:
y_pred = cross_val_predict(BBagging, X, np.asarray(y).reshape(y.shape[0],), cv=8)
AUC = roc_auc_score(y, y_pred)
print("AUC: ",AUC)

AUC:  0.9089635854341738


In [66]:
print("Confusion matrix: \n",confusion_matrix(y,y_pred))

Confusion matrix: 
 [[49  2]
 [ 3 18]]


In [81]:
from imblearn.ensemble import EasyEnsembleClassifier
Ensemble=EasyEnsembleClassifier(base_estimator=GaussianNB(),n_estimators=20,sampling_strategy="not majority",random_state=0)
scores=cross_val_score(Ensemble,X,np.asarray(y).reshape(y.shape[0],),cv=8)
print("Accuracy: ",scores.mean(),"+/-",scores.std())

Accuracy:  0.9305555555555556 +/- 0.07733006059486142


Both BalancedBaggingClassifier and EasyEnsembleClassifier produced average accuracy of 93.06%. However,the BalanacedBaggingClassifier had a smaller standar deviation, implying that the BalancedBaggingClassifier is more stable than EasyEnsembleClassifier.

The BalancedBaggingClassifier model misclassified only 5 instances. The algorithm won't run on only 5 instances. Even if we just do a 2-fold cross validation, one of the folds would probably have only 1 class.(It has already been experimented.) Instead of running a reinforcement round on the instances that the model missed during the initial round, I will try to see if the model's performance can be further improved by adjusting the threshold.

In [82]:
y_pred=pd.DataFrame(cross_val_predict(BBagging, X, np.asarray(y).reshape(y.shape[0],), 
                                      cv=8,method='predict_proba')).rename(columns={0:"pred_0",1:"pred_1"})
y_pred.head()

pred_0    pred_1
0  0.593201  0.406799
1  0.476837  0.523163
2  0.004543  0.995457
3  0.491090  0.508910
4  0.000021  0.999979

In [83]:
y_probs1=y_pred[["pred_1"]]

In [84]:
import numpy as np
thresholds=np.arange(0,1,0.001)

In [85]:
def to_labels(pos_probs,threshold):
    return (pos_probs >= threshold).astype('int')

In [86]:

from sklearn.metrics import f1_score
scores=[f1_score(y, to_labels(y_probs1,t),average="micro") for t in thresholds]

In [87]:
ix=np.argmax(scores)
print((thresholds[ix],scores[ix]))

(0.364, 0.9444444444444444)


In [39]:
len(y_probs1[y_probs1["pred_1"]>=0.367])

21

In [88]:
y_pred["pred_0"]=np.where((y_pred.pred_0 >=0.364),1,y_pred.pred_0)
y_pred["pred_0"]=np.where((y_pred.pred_0 < 0.364),0,y_pred.pred_0)
y_pred["pred_1"]=np.where((y_pred.pred_1 >=0.364),1,y_pred.pred_1)
y_pred["pred_1"]=np.where((y_pred.pred_1 <0.364),0,y_pred.pred_1)

In [89]:
len11=len(y_pred[y_pred["pred_1"]==1])
print("The number of predicted positive cases: ",len11)

The number of predicted positive cases:  21


In [90]:
y_pred["pred_0"]=np.where((y_pred.pred_0 == y_pred.pred_1),0,y_pred.pred_0)

In [91]:
len00=len(y_pred[y_pred["pred_0"]==1])
print("The number of predicted negative cases: ",len00)

The number of predicted negative cases:  51


In [92]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder(handle_unknown="ignore")
y_true=pd.DataFrame(enc.fit_transform(y).toarray())

In [93]:
y_pred=enc.inverse_transform(y_pred)

In [94]:
print("AUC after adjusting for the threshold: ",roc_auc_score(y,y_pred))

AUC after adjusting for the threshold:  0.9327731092436975


In [95]:
print("Confusion_matrix after adjusting the threshold: \n",confusion_matrix(y,y_pred))

Confusion_matrix after adjusting the threshold: 
 [[49  2]
 [ 2 19]]


In [96]:
print("Accuracy after adjusting the threshold: ",(49+19)/(49+2+2+19))

Accuracy after adjusting the threshold:  0.9444444444444444


Adjusting the threshold improved the model's accuracy by 1.4% and AUC by 0.024.